ansible_galaxy/actions/publish.py

```diff
@@ -0,0 +1,89 @@
+
+import hashlib
+import json
+import logging
+import os
+import tarfile
+
+from ansible_galaxy.rest_api import GalaxyAPI
+from ansible_galaxy import collection_artifact_manifest
+from ansible_galaxy import exceptions
+
+log = logging.getLogger(__name__)
+
+
+def _get_file_checksum(file_path):
+    checksum = hashlib.sha256()
+    with open(file_path, "rb") as f:
+        for byte_block in iter(lambda: f.read(4096), b""):
+            checksum.update(byte_block)
+    return checksum.hexdigest()
+
+
+def _publish(galaxy_context,
+             archive_path,
+             display_callback=None):
+
+    results = {
+        'errors': [],
+        'success': True
+    }
+
+    archive = tarfile.open(archive_path, 'r')
+    top_dir = os.path.commonprefix(archive.getnames())
+    manifest_path = os.path.join(top_dir,
+                                 collection_artifact_manifest.COLLECTION_MANIFEST_FILENAME)
+    try:
+        manifest_file = archive.extractfile(manifest_path)
+    except Exception as exc:
```

Comment:
a tarfile.Tarfile should mostly raise tarfile.TarError based errors, so I would just catch those here.